04_snake_game/main.go

```diff
@@ -93,21 +170,96 @@ func mod(n, m int) int {
 
 // Makes a move for a snake. Returns a snake with an updated position.
 func moveSnake(s snake, v coord, fw, fh int) snake {
-	s.pos.x = mod(s.pos.x+v.x, fw)
-	s.pos.y = mod(s.pos.y+v.y, fh)
+	copy(s.pos[1:], s.pos[:])
+	s.pos[0] = coord{s.pos[0].x + v.x, s.pos[0].y + v.y}
 	return s
 }
 
+func HitsTheWalls(g game) bool {
+	str := fmt.Sprintln("Game Over")
+	if g.sn.pos[0].x == 0 || g.sn.pos[0].x == g.fieldWidth-1 || g.sn.pos[0].y == 1 || g.sn.pos[0].y == g.fieldHeight-1 {
+		termbox.Clear(snakeFgColor, snakeBgColor)
+		writeText((g.fieldWidth-len(str))/2, g.fieldHeight/2, str, termbox.ColorRed|termbox.AttrBold, termbox.ColorDefault)
+		termbox.Flush()
+		time.Sleep(5 * time.Second)
+		termbox.Clear(snakeFgColor, snakeBgColor)
+		termbox.Flush()
+		return true
+	}
+	return false
+}
+
+func HitsItself(g game) bool {
+	str := fmt.Sprintln("Game Over")
+	for _, pos := range g.sn.pos[2:] {
+		if g.sn.pos[0] == pos {
+			termbox.Clear(snakeFgColor, snakeBgColor)
+			writeText((g.fieldWidth-len(str))/2, g.fieldHeight/2, str, termbox.ColorRed|termbox.AttrBold, termbox.ColorDefault)
+			termbox.Flush()
+			time.Sleep(5 * time.Second)
+			termbox.Clear(snakeFgColor, snakeBgColor)
+			termbox.Flush()
+			return true
+		}
+	}
+	return false
+}
+
 func step(g game) game {
 	g.sn = moveSnake(g.sn, g.v, g.fieldWidth, g.fieldHeight)
 	draw(g)
```

Comment:
nit: drawing should happen after checks for collisions, otherwise, the snake and an apple overlap.

Reply:
I fixed it.